HPCC-13984 Increase listen backlog for large clusters with many thor slaves

mckellyln opened the pull request:
Signed-off-by: Mark Kelly mark.kelly@lexisnexis.com

mckellyln committed:
HPCC-13984 Increase listen backlog for large clusters with many thor slaves

Added check for kernel somaxconn and env mpSoMaxConn at MP start up.
Added changes to conf and xml file for mpStart, mpEnd, mpSoMaxConn to
not be nested so that configmgr will use with new configs.
Fixed typo in xml.in.
Updated testing/regress/environment.xml.in to match same changes.

Signed-off-by: Mark Kelly <mark.kelly@lexisnexis.com>

mckellyln commented:
@richardkchapman please review if you have time.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented:
The change looks innocuous enough though I wonder who put the comment there and why?

You note in the Jira "NOTE: the kernel tuneable "somaxconn" should also be increased to handle this plus other thors on the same host." - is that something we should be changing in the init scripts/post-install scripts? @Michael-Gardner ?

mckellyln commented:
Don't know about the origin of the comment.
Do we adjust kernel tuneables now at install time ?  If not I would suggest we continue to not adjust (but perhaps suggest what should be modified), but if we do, then we should check the tuneable value and only increase it if its less than what we suggest.

Michael-Gardner commented:
I believe two different settings would need to be changed on the thormaster (not the slave machines.) net.core.somaxconn should be raised to 1024, and we should insure that the number of open file descriptors matches as well (bash built in ulimit.)

mckellyln commented:
needs to increased on all slaves machines also. I would probably go larger than 1024 as well.
Does install adjust kernel tuneables now ?

Michael-Gardner commented:
The place where we deal with tunables is in the init scripts.  There is a section in hpcc_common during startCmd() that runs for all processes, and anything else that needs to be tuned per process is taken care of in that specific processes init script.

mckellyln commented:
We do not adjust kernel tuneables at install time, except perhaps the cassandra plugin.

richardkchapman commented:
I thought we did something for wmem_max/rmem_max for roxie, though I can't see where. We DO check in Roxie that htey are set to adequate values and fail if they are not... perhaps thor should do something similar (or at least warn) ?

mckellyln commented:
ok, I see check_set() in udplib.

mckellyln commented:
@richardkchapman added warning and updated mptest code.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

mckellyln commented on system/mp/mpcomm.cpp:
```diff
@@ -1889,6 +1914,14 @@ int CMPConnectThread::run()
             try {
+                if (soMaxCheck)
+                {
+                    if (numMPAccepts++ >= soMaxConn/2)
```
It may be possible to look at connect() failures from the other side and give a warning as well, but that can be added as another task.
Ultimately the mp connection should launch or pass to another thread to complete the handshake so that accept()s happen very quickly and there is little chance for the backlog to grow.

mckellyln commented on system/mp/mpcomm.cpp:
```diff
@@ -1889,6 +1914,14 @@ int CMPConnectThread::run()
             try {
+                if (soMaxCheck)
+                {
+                    if (numMPAccepts++ >= soMaxConn)
```
Could also maybe check other side for connect() failures and give warning, but that should be a separate task.
Ultimately should launch or pass to another thread to complete the MP handshake so that accept()s happen very quickly to reduce chance of backlog growing large.

richardkchapman replied:
This seems to be counting the total number rather than the simultaneous number - I don'r really know how the mp stuff works but this thread appears to continue running until the client closes down (do nodes reconnect, or connect more than once) which might lead to spurious warnings I think?

mckellyln replied:
ok, for Dali I agree it may just be counting.  I will rework.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented on system/mp/mpcomm.cpp:
```diff
@@ -1801,7 +1823,7 @@ CMPConnectThread::CMPConnectThread(CMPServer *_parent, unsigned port)
             try
             {
-                listensock = ISocket::create(port, 16);  // better not to have *too* many waiting
+                listensock = ISocket::create(port, 800);
```
I assumed you would be checking here (to see if the value you were trying to set was greater than somaxconn).

If you are running on an older kernel the value is hardwired to 128 (and defined by the macro SOMAXCONN) I think.

mckellyln replied:
I will rework to try and account for concurrent connections.  I do not want to warn always for single node and small configs.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

mckellyln commented:
check for environment setting mpSoMaxConn at MP startup and if found compare to kernel somaxconn to warn user if needed.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

mckellyln commented:
Added mpStart, mpEnd, mpSoMaxConn to initial env.conf and xml and also moved these out of nested ports section of xml so that new configs will honor these settings.  Code may read old nested ports section to maintain backward compatibility.  @garonsky please review.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: Fail
HPCC Stop: OK

garonsky commented:
Looks good.

richardkchapman commented on initfiles/etc/DIR_NAME/environment.conf.in:
```diff
@@ -21,7 +21,9 @@ interface=*
 use_epoll=true
 # allow kernel pagecache flushing where enabled (true/false)
 allow_pgcache_flush=true
-
```
Why here as well??

mckellyln replied:
This is so new configs will get these settings.  Configmgr creates the Environment section for new configs from what is in the conf file.  It is in the xml.in so the initial/default xml has these settings, because the initial xml file is not generated from configmgr.

richardkchapman commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -15,10 +15,9 @@
   <runtime>${RUNTIME_PATH}</runtime>
   <sourcedir>${CONFIG_SOURCE_PATH}</sourcedir>
   <user>${RUNTIME_USER}</user>
-  <ports>
```
What was wrong with the old path exactly? Moving it seems like scope for confusion unless there's a good reason

mckellyln replied:
Configmgr does not create a ports section when creating new configs, but will put what is in the conf file into the Environment section (without nesting).  To make it so a hand edit to the xml is not required for this setting with configs generated by configmgr it was decided to flatten the ports subsection.  @garonsky can elaborate.  Another route is to add to configmgr to support nested sections from the conf file but that seems a more involved method.

garonsky replied:
It doesn't appear that CProperties in jlib supports creating nested properties, and that's how that sections is written/read by the config system. Also, https://track.hpccsystems.com/browse/HPCC-12750?jql=text%20~%20%22generate%20environment%22 would completely remove the static environment.xml from the source tree, which would eliminate the nesting entirely from source, but retain backward compatibility.  

If we need hierarchy in the EnvSettings section of the environment.xml we could convert it to XSD based, which probably is the right thing to do long term for every config setting.

richardkchapman commented:
@mckellyln A couple of questions

mckellyln commented on initfiles/etc/DIR_NAME/environment.xml.in:
```diff
@@ -15,10 +15,9 @@
-  </ports>
+  <mpStart>7101</mpStart>
+  <mpEnd>7500</mpEnd>
+  <mpSoMaxConn>128</mpSoMaxConn>
```
Just fixed typo here.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

mckellyln commented:
updated env.xml.in in testing/regress also to match.

HPCCSmoketest commented:
Automated Smoketest Build: success
HPCC Start: OK
HPCC Stop: OK

richardkchapman commented:
I have some reservations about the relationship between environment.xml and environment.conf - seems to me that putting values in environment.conf to be used when you use the wizard to create a new environment.xml, but having the system read the values from environment.xml (and/or from the copy of it that is copied into dali at startup) is liable to lead to confusion, when people edit one and discover that nothing changes. Other settings are read directly from environment.conf, while yet others have their defaults controlled by the xsd file but can be modified in the environment.cml file after it is created...

That's not really a new issue related to this change though, so I will accept this change while raising a Jira to discuss the wider issue later.
